Create and Manage File Upload Exercise

maxr96 opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested the changes and all related features on the test server https://artemistest.ase.in.tum.de.
- [X] I documented my source code using the JavaDoc / JSDoc style.
- [X] I added integration test cases for the server (Spring) related to the features
- [X] I added integration test cases for the client (Jest) related to the features
- [X] I added screenshots/screencast of my UI changes
- [X] I translated all the newly inserted strings

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
Users should be able to create and manage File Upload Exercises.

### Description
<!-- Describe your changes in detail -->
File Upload Exercise is available in Course Administration page. Users can create new exercises, delete, view and edit them.


### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

1. Log in to Artemis
2. Navigate to Course Administration
3. Scroll to File Upload Exercises and click 'Create a new File Upload Exercise'

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->
![Screenshot_2019-08-30 Courses](https://user-images.githubusercontent.com/17262376/64016645-d76e3d80-cb27-11e9-8513-72c7c489a14b.png)
---
![Screenshot_2019-09-01 File Upload Exercises(2)](https://user-images.githubusercontent.com/17262376/64078560-7f2c6c80-ccdc-11e9-9f65-ea027197cba7.png)
---
![Screenshot_2019-09-01 File Upload Exercises(1)](https://user-images.githubusercontent.com/17262376/64078567-8a7f9800-ccdc-11e9-90f6-16d23920e15e.png)
---
![Screenshot_2019-08-30 Courses(1)](https://user-images.githubusercontent.com/17262376/64017387-c7eff400-cb29-11e9-98dd-c0e291bbaa9b.png)

thilo-behnke commented on src/main/webapp/app/entities/file-upload-exercise/file-upload-exercise-detail.component.ts:
```diff
@@ -20,15 +21,18 @@ export class FileUploadExerciseDetailComponent implements OnInit, OnDestroy {
+        this.fileUploadExerciseService
+            .find(exerciseId)
+            .pipe(filter(res => !!res.body))
+            .subscribe((fileUploadExerciseResponse: HttpResponse<FileUploadExercise>) => {
```
It would be nice to have some error handling here.

maxr96 replied:
Good suggestion. Added.

thilo-behnke commented on src/main/webapp/app/entities/file-upload-exercise/file-upload-exercise-update.component.html:
```diff
@@ -1,13 +1,100 @@
     <div class="col-8">
         <form name="editForm" role="form" novalidate (ngSubmit)="save()" #editForm="ngForm">
-            <h2 id="jhi-file-upload-exercise-heading" jhiTranslate="artemisApp.fileUploadExercise.home.createOrEditLabel">Create or edit a File Upload Exercise</h2>
+            <h2 *ngIf="fileUploadExercise.id" id="jhi-file-upload-exercise-heading-edit">{{ 'artemisApp.fileUploadExercise.home.editLabel' | translate }}</h2>
```
You can also use `jhiTranslate="artemisApp..." on the tag.

maxr96 replied:
Do you think this is a better approach? In different places we use different approaches. I personally, prefer the one I use, as it is shorter.

maxr96 replied:
After some clarification I decided to use jhiTranslate here

thilo-behnke commented on src/main/webapp/app/entities/file-upload-exercise/file-upload-exercise.component.ts:
```diff
@@ -1,56 +1,70 @@
+/* angular */
```
We don't use this kind of comment anywhere atm.
We could add linting rules if we want to enforce a specific order.

maxr96 replied:
Yes, you are right. I deleted it

thilo-behnke commented on src/main/webapp/app/entities/file-upload-exercise/file-upload-exercise.module.ts:
```diff
@@ -7,39 +7,32 @@ import { ArtemisSharedModule } from 'app/shared';
+        ArtemisMarkdownEditorModule,
     ],
+    declarations: [FileUploadExerciseComponent, FileUploadExerciseDetailComponent, FileUploadExerciseUpdateComponent, FileUploadExerciseDeleteDialogComponent],
+    entryComponents: [FileUploadExerciseComponent, FileUploadExerciseUpdateComponent, FileUploadExerciseDeleteDialogComponent],
```
Please remove all components here, that are not opened in a dialog anymore.

maxr96 replied:
Done

maxr96 commented:
> 
> 
> To me this looks good in general, only some minor issues.
> 
> I also found the following issues when testing:
> 
>     * Edit > save & Edit > back (browser) don't lead to Course Administration > Exercises but all the way back to Course Administration
> 
>     * I don't understand what the file pattern property means:
>       -> What does the setting do?
>       -> Why is it not mandatory? What is the behavior when no file pattern is entered?
>       -> We should probably have a tooltip here.

Thank you for the feedback. I resolved the issues you pointed. 

- I fixed the 'Cancel' button for Edit File Upload Exercise. Since we navigate there only from course page, I simply navigate back to that route.
- File Pattern is used to specify which file extensions can be used to submit exercise. It is a comma separated list. For now it will be a simple input field, but later on we can make it as a checkbox. It is required now and I added default formats: 'pdf, png'. Also I added an info box and updated the screenshots.

maxr96 commented:
> 
> 
> Cool, two issues before I can approve it:
> 
>     1. Bug: on opening the edit screen, the existing file pattern settings are overwritten with `pdf, png`.
> 
>     2. Consistency: I would rather use a question mark icon with a tooltip next to the label of the file pattern option as we normally use the Bootstrap alerts for displaying error/success messages (see e.g. the programming-exercise creation screen)

1. Yes, you are right here, missed it.
2. It is a very good point. I first thought about the info icon that we use, while creating complaints and it looks different from the one used in process of creating Programming Exercise. However, I chose icon used for Programming Exercise as in my opinion it is better to have question mark instead of exclamation mark and I prefer to use `ngbTooltip` instead of native HTML tooltip. 
It can be a good idea to check other places, where we use icons for showing additional information.

lovaiible commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadExerciseResource.java:
```diff
@@ -152,17 +150,17 @@ public FileUploadExerciseResource(FileUploadExerciseRepository fileUploadExercis
+        Optional<FileUploadExercise> fileUploadExercise = fileUploadExerciseRepository.findById(exerciseId);
         Course course = fileUploadExercise.get().getCourse();
         User user = userService.getUserWithGroupsAndAuthorities();
         if (!authCheckService.isTeachingAssistantInCourse(course, user) && !authCheckService.isInstructorInCourse(course, user) && !authCheckService.isAdmin()) {
```
You can just use the method `isAtLeastTeachingAssistantInCourse(course, user)` here instead. Also there is the method `isAtLeastInstructorInCourse(course, user)` which can be used for line 85 e.g.

maxr96 replied:
Good point, makes it easier to read code

lovaiible commented on src/main/resources/config/application-artemis.yml:
```diff
@@ -12,17 +12,17 @@ artemis:
         url: https://jirabruegge.in.tum.de
-        user: <username>    # e.g. ga12abc
-        password: <password>
+        user: ge56miw    # e.g. ga12abc
```
I think you accidentally pushed the changes here. Can you revert them?

maxr96 replied:
True, silly mistake

lovaiible commented on src/main/resources/config/application-dev.yml:
```diff
@@ -34,7 +34,7 @@ spring:
         url: jdbc:mysql://localhost:3306/Artemis?createDatabaseIfNotExist=true&useUnicode=true&characterEncoding=utf8&useSSL=false&useLegacyDatetimeCode=false&serverTimezone=UTC
         username: root
-        password:
+        password: rootpwd
```
> I think you accidentally pushed the changes here. Can you revert them?
Same here

lovaiible commented on src/main/webapp/app/entities/file-upload-exercise/file-upload-exercise-delete-dialog.component.html:
```diff
@@ -8,11 +8,17 @@ <h4 class="modal-title" jhiTranslate="entity.delete.title">Confirm delete operat
             Are you sure you want to delete this File Upload Exercise?
         </p>
+        <div class="form-group">
+            <label class="text-danger">Please type in the name of the Exercise to confirm.</label>
```
The translation for this text is missing

maxr96 replied:
Yes, I added it now

lovaiible commented on src/main/webapp/app/entities/file-upload-exercise/file-upload-exercise.component.ts:
```diff
@@ -1,56 +1,74 @@
+    /**
+     * Returns the unique identifier for items in the collection
+     * @param index of a file upload exercise in the collection
+     * @param item current user
```
I think that's a copy and paste error here :)

lovaiible commented on src/main/webapp/app/entities/file-upload-exercise/file-upload-exercise.route.ts:
```diff
@@ -26,54 +63,45 @@ export const fileUploadExerciseRoute: Routes = [
     {
-        path: 'file-upload-exercise/:id/delete',
-        component: FileUploadExerciseDeletePopupComponent,
+        path: 'course/:courseId/file-upload-exercise/:id',
```
Shouldn't it be `course/:courseId/file-upload-exercise/:exerciseId` to be consistent?

maxr96 replied:
Yes, good point

lovaiible commented on src/main/webapp/app/entities/file-upload-exercise/file-upload-exercise.route.ts:
```diff
@@ -26,54 +63,45 @@ export const fileUploadExerciseRoute: Routes = [
-        path: 'course/:courseId/file-upload-exercise/:id',
-        component: FileUploadExerciseDetailComponent,
         data: {
             authorities: ['ROLE_TA', 'ROLE_INSTRUCTOR', 'ROLE_ADMIN'],
```
I'm a bit confused. In the `FileUploadExerciseResource` you only authorized Instructors and Admins to create and update File Upload Exercises, but here the path is accessible for TAs as well? Is it intended this way?

maxr96 replied:
You are right here I changed the authorities

lovaiible commented on src/main/webapp/i18n/de/fileUploadExercise.json:
```diff
@@ -2,20 +2,25 @@
+            "deleted": "Datei Upload Aufgabe gelöscht mit ID {{ param }}",
             "delete": {
-                "question": "Soll File Upload Exercise {{ title }} wirklich dauerhaft gelöscht werden?"
+                "question": "Soll Datei Upload Aufgabe{{ title }} wirklich dauerhaft gelöscht werden?"
```
```suggestion
                "question": "Soll die Datei Upload Aufgabe{{ title }} wirklich dauerhaft gelöscht werden?"
```

lovaiible commented on src/main/webapp/i18n/de/fileUploadExercise.json:
```diff
@@ -2,20 +2,25 @@
-            "filePattern": "File Pattern"
+            "filePattern": "Dateimuster",
+            "filePatternInfo": "Hier kannst du eine kommagetrennte Liste der erlaubten Dateierweiterungen angeben.",
+            "submitSuccessful": "Deine Antwort wurde erfolgreich abgesendet! Du kannst dein Antwort nun nicht mehr bearbeiten. Bitte warte auf dein Feedback.",
```
```suggestion
            "submitSuccessful": "Deine Antwort wurde erfolgreich abgesendet! Du kannst deine Antwort nun nicht mehr bearbeiten. Bitte warte auf dein Feedback.",
```

lovaiible commented on src/main/webapp/i18n/de/fileUploadExercise.json:
```diff
@@ -2,20 +2,25 @@
+            "filePattern": "Dateimuster",
+            "filePatternInfo": "Hier kannst du eine kommagetrennte Liste der erlaubten Dateierweiterungen angeben.",
+            "submitSuccessful": "Deine Antwort wurde erfolgreich abgesendet! Du kannst dein Antwort nun nicht mehr bearbeiten. Bitte warte auf dein Feedback.",
+            "submitDeadlineMissed": "Deine Antwort wurde erfolgreich abgesendet! Du kannst dein Antwort nun nicht mehr bearbeiten. Bitte warte auf dein Feedback. Da du die Deadline verpasst hast wird deine Bewertung nicht gezählt.",
```
```suggestion
            "submitDeadlineMissed": "Deine Antwort wurde erfolgreich abgesendet! Du kannst deine Antwort nun nicht mehr bearbeiten. Bitte warte auf dein Feedback. Da du die Deadline verpasst hast, wird deine Bewertung nicht gezählt.",
```

lovaiible commented on src/main/webapp/i18n/de/fileUploadExercise.json:
```diff
@@ -2,20 +2,26 @@
             "delete": {
-                "question": "Soll File Upload Exercise {{ title }} wirklich dauerhaft gelöscht werden?"
+                "question": "Soll die Datei Upload Aufgabe {{ title }} wirklich dauerhaft gelöscht werden?",
+                "typeNameToConfirm": "Bitte gibst du den Namen der Übung zur Bestätigung ein."
```
```suggestion
                "typeNameToConfirm": "Bitte gebe den Namen der Übung zur Bestätigung ein."
```
